captum/attr/_core/occlusion.py

```diff
@@ -0,0 +1,282 @@
+                            device contain at most
+                            (ablations_per_eval * #examples) / num_devices
+                            samples.
+                            Default: 1
+
+        Returns:
+                *tensor* or tuple of *tensors* of **attributions**:
+                - **attributions** (*tensor* or tuple of *tensors*):
+                            The attributions with respect to each input feature.
+                            Attributions will always be
+                            the same size as the provided inputs, with each value
+                            providing the attribution of the corresponding input index.
+                            If a single tensor is provided as inputs, a single tensor is
+                            returned. If a tuple is provided for inputs, a tuple of
+                            corresponding sized tensors is returned.
+
+
+        Examples::
+
+            >>> # SimpleClassifier takes a single input tensor of size Nx4x4,
+            >>> # and returns an Nx3 tensor of class probabilities.
+            >>> net = SimpleClassifier()
+            >>> # Generating random input with size 2 x 4 x 4
+            >>> input = torch.randn(2, 4, 4)
+            >>> # Defining FeatureAblation interpreter
+            >>> ablator = Occlusion(net)
+            >>> # Computes occlusion attribution, ablating each 1x3x3 patch,
+            >>> # shifting in each direction by the default of 1.
+            >>> attr = ablator.attribute(input, target=1, occlusion_shapes=(1,3,3))
+        """
+        formatted_inputs = _format_input(inputs)
+
+        # Formatting strides
+        if strides is None:
+            strides = tuple(1 for input in formatted_inputs)
+        if len(formatted_inputs) == 1 and not (
+            isinstance(strides, tuple) and len(strides) == 1
+        ):
+            strides = (strides,)
+        assert isinstance(strides, tuple) and len(strides) == len(
+            formatted_inputs
+        ), "Strides must be provided for each input tensor."
+
+        # Construct occlusion blocks
+        if not isinstance(occlusion_shapes[0], tuple):
+            occlusion_shapes = (occlusion_shapes,)
+        assert len(occlusion_shapes) == len(
+            formatted_inputs
+        ), "Must provide occlusion dimensions for each tensor."
+        occlusion_tensors = tuple(
+            torch.ones(occ_shape, device=formatted_inputs[i].device)
+            for i, occ_shape in enumerate(occlusion_shapes)
+        )
+
+        # Construct counts, defining number of steps to make of occlusion block in
+        # each dimension.
+        shift_counts = []
+        for i, inp in enumerate(formatted_inputs):
+            current_shape = np.subtract(inp.shape[1:], occlusion_shapes[i])
+            shift_counts.append(
+                tuple(
+                    np.add(np.ceil(np.divide(current_shape, strides[i])).astype(int), 1)
+                )
+            )
```

Comment:
Good point, I think we should probably assert and raise an exception in this case. Even if the shift count was 1, it wouldn't cover the entire second dimension, so we would have missing attributions.

Reply:
Yeah, since it is always doing plus one we might need to check that too. For example in this case:
```
formatted_inputs = (torch.rand(1, 3, 255, 255), )
occlusion_shapes = ((2, 4, 4), )
strides = (3,)
```
I'm still getting `[(2, 85, 85)]`

Reply:
Yes, so in this case, I intended for it to return 2 for the first dimension to ensure that it covers the full range. This results in the final patch / occlusion block to be cut off, but includes the entire range. I have a note about this in the documentation.

Reply:
Sorry, first misread the values. This is the same case as above right? With a stride of 3, index 2 of dimension 1 is still not covered?

Reply:
yes, it is similar case: the only difference is that the stride length is equal to the 1st dimension(count starting from 0) of input. since stride length is 3, I expect to get: `[(1, 85, 85)]` instead of `[(2, 85, 85)]`, is that right ?

Reply:
Also, shouldn't `shift_counts` be `shift_counts=(1,33,33)` ?

Reply:
Hm, I think the 85, 85, is correct? The 85 makes sense since stride is 3, so last shift starts at 84*3 = 252. The extra 1 counts the origin transformation. The first dimension is off since the stride is >= (input dimension - occlusion dimension), which causes skipped regions. Will detect and raise an exception in that case, so hopefully should resolve this issue!

Reply:
Yes, 85, 85 are correct. My point was for 2 in (2, 85, 85)

Reply:
Oh ok, makes sense, sounds good!